### PR TITLE
Add flex voting support to the bravo governor

### PR DESCRIPTION
**Pull request by davidlaprade**
⚠️ This should not merge until https://github.com/ScopeLift/flexible-voting/pull/31 hits master ⚠️

### CODE REVIEW

File: test/GitcoinGovernor.t.sol
Comment (mds1) [inline test/GitcoinGovernor.t.sol]: Can we change this to `forge install transmissions11/solmate@v7` and then just import the library, or just import the version we already have in `lib/flexible-voting/lib/solmate`? If there's some reason we don't want the dep, let's put this in it's own file in `test/lib` and import it. Sticking a library in the middle of a bunch of tests isn't too conventional and makes it hard to find/know about.

(Also in general when you have a comment like `// Copied from Solmate`, best practice is include a permalink (i.e. URL includes the commit hash) to the original source)

File: test/GitcoinGovernor.t.sol
Comment (mds1) [inline test/GitcoinGovernor.t.sol:1119]: What's the testing philosophy you took here? Copy a subset of tests from the flexible voting around core functionality and modify them as needed? Some alternative ideas include:
- A single simple test to verify we are inheriting from the fractional governor extension, e.g. the `assertEq(governorBravo.COUNTING_MODE(), "support=bravo&quorum=bravo&params=fractional")` line above. The argument here is: we know the fractional governor works, just check that we are inheriting from it
- Copy the full fractional test suite into it's own file and change the contract it runs against to be the gitcoin governor, to ensure we didn't accidentally break any functionality. This is what I typically do when inheriting from a contract/repo where it's easy to do this

Also partly asking so I know how closely to review all these tests. If you wrote them all from scratch/memory I should review carefully, if you copied them with no or minimal changes (if minimal, what changes) my review can be much quicker

**Review by mds1 — COMMENTED** [review]
LGTM!

File: test/GitcoinGovernor.t.sol
Comment (apbendi) [reply test/GitcoinGovernor.t.sol:1119]: Looks like these tests aren't copy/pasta'd from the Fractional repo because they are testing the Governor in situ after the upgrade against the fork. In other words, it's testing that fractional voting works post upgrade, with the GitcoinGovernor assembled according to Gitcoin's requirements, which is not what we do in the Fractional Governor unit tests. In fact that's the point of all these tests, as the Governor components themselves are in fact tested by OZ.

File: test/GitcoinGovernor.t.sol
Comment (apbendi) [inline test/GitcoinGovernor.t.sol:652]: Definitely don't feel strongly enough to hold up the PR or require a change, but just want to note that in general I don't love moving the assertions into the setup helpers. I think it's nice to have the nature of the assertions obvious in each test, even if it's duplicated. If I do extract an assertion I try to make that a separate method from setup so it's clear what's going on, so it would look something like.

```solidity
_submitTokenSendProposalToGovernorBravo(address(gtcToken), _gtcAmount, _gtcReceiver);
_assertPending(_newProposalId);
```

Again, no change required, just wanted to share a perspective on test clarity.

**Review by apbendi — APPROVED** [review]
Looks good to me! Glad we didn't encounter any unexpected issues.

File: test/GitcoinGovernor.t.sol
Comment (mds1) [reply test/GitcoinGovernor.t.sol:1119]: > In other words, it's testing that fractional voting works post upgrade, with the GitcoinGovernor assembled according to Gitcoin's requirements, which is not what we do in the Fractional Governor unit tests.

So IIUC we're basically testing that the fractional voting logic wasn't overridden accidentally when assembling the governor?

File: test/GitcoinGovernor.t.sol
Comment (mds1) [reply test/GitcoinGovernor.t.sol:652]: Alternatively I sometimes like just renaming the function if it holds business-logic assertions, like `_submitTokenSendProposalToGovernorBravoAndAssertXYZ`. When the assertions are things like "making sure the storage overwrite was done successfully" then that would not be reflected in the var name / seems ok to hide in a helper method

File: test/GitcoinGovernor.t.sol
Comment (davidlaprade) [reply test/GitcoinGovernor.t.sol]: Yeah, good points all around.

I think for some reason I didn't want to import solmate from within the flex voting lib, but now I'm not really sure why. I'll just do that

File: test/GitcoinGovernor.t.sol
Comment (davidlaprade) [reply test/GitcoinGovernor.t.sol:1119]: > What's the testing philosophy you took here?

Good question. I wrote new tests to cover the flex voting features that I thought Gitcoin would care about having, as well as some sanity tests that those features wouldn't interfere with normal governor functioning. Otherwise I didn't retest `GovernorCountingFractional`, since we've already done that

File: test/GitcoinGovernor.t.sol
Comment (davidlaprade) [reply test/GitcoinGovernor.t.sol:652]: Yeah, I generally do prefer just having explicitness in tests, even if it means code duplication. But the duplication had been called out on a previous PR so I was trying to cut it down a bit here

File: test/GitcoinGovernor.t.sol
Comment (davidlaprade) [reply test/GitcoinGovernor.t.sol]: Done in 14ae0bf. I ended up just adding the dependency directly

> I think for some reason I didn't want to import solmate from within the flex voting lib

I went to import from within flex-voting and then remembered why I didn't do it before: it was going to require a remapping, which felt like a major hack. But it also just felt gratuitous to add that whole dependency just for a test file, which is why I decided to copy in the small portion we needed.

Comment (mds1) [reply test/GitcoinGovernor.t.sol]: Awesome, thanks! IMO remappings are always ok, definitely not a major hack—they're pretty much always present and often just behind the scenes anyway, so solc knows where to find stuff

File: test/GitcoinGovernor.t.sol
Comment (mds1) [reply test/GitcoinGovernor.t.sol:1119]: Thanks, appreciate the explanation!

File: test/GitcoinGovernor.t.sol
Comment (mds1) [inline test/GitcoinGovernor.t.sol:1303]: Is there a method that checks if quorum is reached? Seems we're verifying that manually but we might change quorum so want to just make sure we don't forget about this at that time. Just mentioning this part in #18 is sufficient. Feel free to resolve this comment yourself.

Comment (davidlaprade) [reply test/GitcoinGovernor.t.sol:1303]: Good call out. Added a note to #18 about this. Thanks!

**Review by mds1 — APPROVED** [review]

**Pull request merged**
